### PR TITLE
Fp16 support for Conformer

### PR DESCRIPTION
**Pull request by bmwshop**
# What does this PR do ?
enables support for fp16 training

Add a one line overview of what this PR aims to accomplish.
enables support for fp16 training by forcing MHSA to float32 or bfloat16 if available.

**Collection**: [Note which collection this PR will affect]
ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below
trainer.precision=16

```python
# Add a code snippet demonstrating how to use this 
```
n/a

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [X ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/utils/precision_utils.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            return True
+        else:
+            return False
+    else:
+        return False
+
+def get_current_precision():
+    """ Determine current precision set by the trainer
+    """
+    if torch.cuda.is_available():
```
Comment (titu1994) [inline nemo/utils/precision_utils.py:33]: Very poor check. Isn't there some other way ? And also depends on when it is called - prior to init, after model init, after PTL init, during ptl training all give different results here c

**Review by titu1994 — CHANGES_REQUESTED** [review]

File: nemo/utils/precision_utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import torch
+
+
+def is_bfloat16_available():
```
Comment (artbataev) [inline nemo/utils/precision_utils.py:18]: Why not to use `torch.cuda.is_bf16_supported()`?

Comment (titu1994) [reply nemo/utils/precision_utils.py:18]: Because only the most recent pytorch supports that

Comment (artbataev) [reply nemo/utils/precision_utils.py:18]: Oh, I see, thanks!

Comment (artbataev) [reply nemo/utils/precision_utils.py:18]: Do we support pytorch 1.9 and earlier? It was added in [1.10.0](https://github.com/pytorch/pytorch/releases/tag/v1.10.0)

Comment (titu1994) [reply nemo/utils/precision_utils.py:18]: We don't, min is 1.11. Guess we can use that then. Dima is looking into a more proper check anyway

**Review by titu1994 — DISMISSED** [review]
Thanks! This looks great !

**Comment by piraka9011** [comment]
Does this resolve the training stability issues?
If so, is there any reason this isn't merged yet or if I could just cherry pick the commit to test in the meantime?

**Comment by bmwshop** [comment]
> Does this resolve the training stability issues? If so, is there any reason this isn't merged yet or if I could just cherry pick the commit to test in the meantime?

sorry need a little more time to double check for issues; will provide an update shortly

**Review by titu1994 — APPROVED** [review]

**Review by titu1994 — DISMISSED** [review]

**Review by VahidooX — APPROVED** [review]
LGTM!

**Pull request merged**